independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationsTransformer.java

```diff
@@ -238,23 +349,44 @@ public final Builder whenContainsNone(Class<? extends Annotation>... annotationN
          * @param predicate
          * @return self
          */
-        public Builder when(Predicate<TransformationContext> when) {
+        public THIS when(Predicate<TransformationContext> when) {
             if (predicate == null) {
                 predicate = when;
             } else {
                 predicate = predicate.and(when);
             }
-            return this;
+            return self();
+        }
+
+        /**
+         * The transformation logic is performed only if all conditions are met.
+         * <p>
+         * Unlike in {@link #transform(Consumer)} the transformation is always applied if conditions are met, i.e.
+         * {@link Transformation#done()} is called automatically.
+         *
+         * @param consumer
+         * @return a new annotation transformer
+         */
+        public AnnotationsTransformer doTransform(Consumer<Transformation> consumer) {
```

Comment:
I'm not so sure about the method name. Maybe we should use something like `thenTransform()` or `applyTransformation()`. Maybe we should rename the original `transform()` method to something else (`build()`?) - this would be a breaking change but that's acceptable in the context of quarkus 3.

And yes, I know that a terminal operation for builders should be always `build()`... but sometimes it's more readable like this ;-).

Reply:
I think just `apply()` would be fine, but this is OK too.

Reply:
The difference between this method and `transform()` is unclear though. This method is probably preferrable. Is there a case for ever using `transform()`? I wonder if we could/should deprecate it.

Reply:
Hmm, at least one reason to use `transform()` is when you want to access something from the transformation context. So I guess we need both. It's probably hard to guide users to this method for the simple cases and to the other one for more complex cases :shrug:

Reply:
> Is there a case for ever using transform()? 
> Hmm, at least one reason to use transform() is when you want to access something from the transformation context. So I guess we need both. 

Yes, in theory you might need to access the annotation target or the "current" annotations. So we need both.

Reply:
I've renamed `doTransform()` to `thenTransform()`...

Reply:
It would be nice if this method was named `transformXYZ`, so that it's close to `transform` when lexicographically sorted, but I couldn't figure out a good value of `XYZ` :shrug:

Reply:
Or rename the original `transform()` to `build()`/`transformationContext()`/??? and use `transform()` for the new method...

Reply:
Maybe `transforming()`?

Reply:
Ok, let's stick with `thenTransform()`...